security-analysis/security-analysis-api/src/main/java/com/powsybl/security/SecurityAnalysisImpl.java

```diff
@@ -69,70 +97,120 @@ public boolean removeInterceptor(SecurityAnalysisInterceptor interceptor) {
+        return CompletableFuture.completedFuture(null).thenCompose(aaVoid -> {
+            String postContStateId;
+            try {
+                postContStateId = queue.take();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new UncheckedInterruptedException(e);
+            }
+            return runOneLoadFlowAsync(workingStateId, postContStateId, postContParameters, resultBuilder, contingency)
+                    .whenComplete((aVoid, throwable) -> queue.add(postContStateId));
+        });
+    }
 
-                                        resultBuilder.contingency(contingency)
-                                                .setComputationOk(lfResult.isOk());
-                                        violationDetector.checkAll(contingency, network, resultBuilder::addViolation);
-                                        resultBuilder.endContingency();
-                                        network.getVariantManager().removeVariant(postContStateId);
+    private CompletableFuture<Void> runOneLoadFlowAsync(String workingStateId,
+            String postContStateId, LoadFlowParameters postContParameters,
+            SecurityAnalysisResultBuilder resultBuilder, Contingency contingency) {
+        return CompletableFuture
+                .runAsync(() -> applyContingency(workingStateId, postContStateId,
+                        contingency), computationManager.getExecutor())
+                .thenCompose(aVoid -> LoadFlow.runAsync(network, postContStateId,
+                        computationManager, postContParameters))
+                .thenApplyAsync(lfResult -> {
+                    setContingencyOkAndCheckViolations(postContStateId, resultBuilder,
+                            contingency, lfResult);
+                    return null;
+                }, computationManager.getExecutor());
+    }
 
-                                        return null;
-                                    }, computationManager.getExecutor());
-                        }
-                    } else {
-                        resultBuilder.preContingency()
-                                .setComputationOk(false)
-                                .endPreContingency()
-                                .build();
-                        futures = new CompletableFuture[0];
-                    }
+    private void setContingencyOkAndCheckViolations(String postContStateId,
+            SecurityAnalysisResultBuilder resultBuilder, Contingency contingency,
+            LoadFlowResult lfResult) {
+        network.getVariantManager().setWorkingVariant(postContStateId);
+        synchronized (resultBuilder) {
+            resultBuilder.contingency(contingency)
+                    .setComputationOk(lfResult.isOk());
+            violationDetector.checkAll(contingency, network, resultBuilder::addViolation);
+            resultBuilder.endContingency();
+        }
```

Comment:
I'm not sure this is a really good API design. The contingency method should create a PostContingencyResultBuilder that is thread safe. The endContingency should add the PostContingencyResult to the SAResult.